packages/remark-images-download/src/index.js

```diff
@@ -2,212 +2,330 @@ const fileType = require('file-type')
-  fs.stat(path, (err, stats) => {
-    if (err) {
-      return fs.mkdir(path, (err) => {
-        if (err) reject(new Error(`Failed to create dir ${path}`))
-        resolve()
-      })
-    }
-
-    if (!stats.isDirectory()) {
-      reject(new Error(`${path} is not a directory!`))
-    }
-
+  fs.mkdir(path, (err) => {
+    if (err) reject(new Error(`Failed to create dir ${path}`))
     resolve()
   })
 })
 
-const checkAndCopy = (from, to) =>
-  new Promise((resolve, reject) => {
-    fs.readFile(from, (err, data) => {
-      if (err) reject(err)
-      if (!data) {
-        return reject(new Error(`Empty file: ${from}`))
+const checkFileType = (name, data) => {
+  if (!data.length) {
+    throw new Error(`Empty file: ${name}`)
+  }
+
+  const type = fileType(data) || {mime: ''}
+  if (!type.mime || type.mime === 'application/xml') {
+    if (!isSvg(data.toString())) {
+      throw new Error(`Could not detect ${name} mime type, not SVG either`)
+    }
+  } else if (type.mime.slice(0, 6) !== 'image/') {
+    throw new Error(
+      `Detected mime of local file '${name}' is not an image/ type`)
+  }
+}
+
+// Creates a Transform stream which raises an error if the file type
+// is wrong or if the file is not a image.
+const makeValidatorStream = (fileName, maxSize) => {
+  let firstChunk = true
+  let totalSize = 0
+
+  return new Transform({
+    flush (cb) {
+      if (totalSize === 0) {
+        cb(new Error(`File at ${fileName} is empty`))
+        return
+      }
+      cb(null)
+    },
+
+    transform (chunk, encoding, cb) {
+      totalSize += chunk.length
+
+      if (maxSize && maxSize < totalSize) {
+        cb(new Error(`File at ${fileName} weighs more than ${maxSize}`))
+        return
       }
-      const type = fileType(data) || {mime: ''}
-      if (!type.mime || type.mime === 'application/xml') {
-        if (!isSvg(data)) {
-          return reject(new Error(`Could not detect ${from} mime type, not SVG either`))
+
+      if (firstChunk) {
+        try {
+          checkFileType(fileName, chunk)
+        } catch (error) {
+          cb(error)
+          return
         }
-      } else if (type.mime.slice(0, 6) !== 'image/') {
-        return reject(new Error(
-          `Detected mime of local file '${from}' is not an image/ type`))
       }
-      fs.copyFile(from, to, (err) => {
-        if (err) return reject(new Error(`Failed to copy ${from} to ${to}`))
 
-        resolve()
-      })
-    })
+      firstChunk = false
+
+      cb(null, chunk)
+    },
   })
+}
+
+const checkAndCopy = async (from, to) => {
+  const data = await promisify(fs.readFile)(from)
+  checkFileType(from, data)
+  try {
+    await promisify(fs.copyFile)(from, to)
+  } catch (err) {
+    throw new Error(`Failed to copy ${from} to ${to}`)
+  }
+}
 
 function plugin ({
   disabled = false,
   maxFileSize = 1000000,
   dirSizeLimit = 10000000,
   downloadDestination = '/tmp',
   localUrlToLocalPath,
+  httpRequestTimeout = 2000, // in milliseconds
```

Comment:
In fact I’m not sure if it’s a good default, and I’m not even sure if this timeout is a good idea if there are often dozens of images in markdown source.